stl/inc/algorithm

```diff
@@ -3836,6 +3849,93 @@ _CONSTEXPR20 void push_heap(_RanIt _First, _RanIt _Last) {
 
+#ifdef __cpp_lib_concepts
+namespace ranges {
+    // VARIABLE ranges::push_heap
+    template <class _It, class _Pr, class _Pj>
+    constexpr void _Push_heap_by_index(const _It _First, iter_difference_t<_It> _Hole,
+        const iter_difference_t<_It> _Top, iter_value_t<_It>&& _Val, _Pr _Pred, _Pj _Proj) {
+        // percolate _Hole to _Top or where _Val belongs, using _Pred
+        _STL_INTERNAL_STATIC_ASSERT(random_access_iterator<_It>);
+        _STL_INTERNAL_STATIC_ASSERT(sortable<_It, _Pr, _Pj>);
+
+        while (_Top < _Hole) {
+            const auto _Idx = static_cast<iter_difference_t<_It>>((_Hole - 1) >> 1); // shift for codegen
+            if (!_STD invoke(_Pred, _STD invoke(_Proj, *(_First + _Idx)), _STD invoke(_Proj, _Val))) {
+                break;
+            }
+
+            // move _Hole up to parent
+            *(_First + _Hole) = _RANGES iter_move(_First + _Idx);
+            _Hole             = _Idx;
+        }
+
+        *(_First + _Hole) = _STD move(_Val); // drop _Val into final hole
+    }
+
+    class _Push_heap_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <random_access_iterator _It, sentinel_for<_It> _Se, class _Pr = ranges::less, class _Pj = identity>
+            requires sortable<_It, _Pr, _Pj>
+        constexpr _It operator()(_It _First, _Se _Last, _Pr _Pred = {}, _Pj _Proj = {}) const {
+            _Adl_verify_range(_First, _Last);
+            auto _UFirst = _Get_unwrapped(_STD move(_First));
+            auto _ULast  = _Get_unwrapped(_STD move(_Last));
+            if constexpr (is_same_v<_It, _Se>) {
+                _Seek_wrapped(_First, _ULast);
+                _Push_heap_unchecked(_STD move(_UFirst), _STD move(_ULast), _Pass_fn(_Pred), _Pass_fn(_Proj));
+            } else if constexpr (sized_sentinel_for<_Se, _It>) {
+                auto _UFinal = _UFirst + (_ULast - _UFirst);
+                _Seek_wrapped(_First, _UFinal);
+                _Push_heap_unchecked(_STD move(_UFirst), _STD move(_UFinal), _Pass_fn(_Pred), _Pass_fn(_Proj));
+            } else {
+                auto _UFinal = _RANGES next(_UFirst, _STD move(_ULast));
+                _Seek_wrapped(_First, _UFinal);
+                _Push_heap_unchecked(_STD move(_UFirst), _STD move(_UFinal), _Pass_fn(_Pred), _Pass_fn(_Proj));
+            }
+
+            return _First;
+        }
+
+        template <random_access_range _Rng, class _Pr = ranges::less, class _Pj = identity>
+            requires sortable<iterator_t<_Rng>, _Pr, _Pj>
+        constexpr borrowed_iterator_t<_Rng> operator()(_Rng&& _Range, _Pr _Pred = {}, _Pj _Proj = {}) const {
+            if constexpr (common_range<_Rng>) {
+                _Push_heap_unchecked(_Ubegin(_Range), _Uend(_Range), _Pass_fn(_Pred), _Pass_fn(_Proj));
+                return _RANGES end(_Range);
+            } else {
+                auto _ULast = _Get_final_iterator_unwrapped(_Range);

```

Comment:
I believe these two eines are the same modulo the function call. 

Also `_First` is an unfortunate name

Reply:
> would unifying the different clauses be worth it?

For the time being I'd like to keep them separate, despite that the former is purely an optimization of the latter. Fewer locals and function calls seems to be a significant improvement in the size of the generated code in debug compiles, even if the difference disappears in release compiles with simple contiguous ranges whose unwrapped iterators are pointers. That said, I don't have the time to analyze these the way I should to make a perf argument, so if folks complain I'll just go the minimally correct route.

> Also _First is an unfortunate name

I'll change these to `_Result`.